File: app/src/main/java/com/epam/valkaryne/openmovieapp/vm/ViewModelFactory.java
package com.epam.valkaryne.openmovieapp.vm;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.epam.valkaryne.openmovieapp.domain.usecase.ClearQueryHistoryUseCase;
import com.epam.valkaryne.openmovieapp.domain.usecase.GetQueryHistoryUseCase;
import com.epam.valkaryne.openmovieapp.domain.usecase.SaveQueryHistoryUseCase;
import com.epam.valkaryne.openmovieapp.domain.usecase.SearchMoviesUseCase;

public class ViewModelFactory implements ViewModelProvider.Factory {

    private final SearchMoviesUseCase searchMoviesUseCase;
    private final GetQueryHistoryUseCase getQueryHistoryUseCase;
    private final SaveQueryHistoryUseCase saveQueryHistoryUseCase;
    private final ClearQueryHistoryUseCase clearQueryHistoryUseCase;

    public ViewModelFactory(SearchMoviesUseCase searchMoviesUseCase,
                            GetQueryHistoryUseCase getQueryHistoryUseCase,
                            SaveQueryHistoryUseCase saveQueryHistoryUseCase,
                            ClearQueryHistoryUseCase clearQueryHistoryUseCase) {
        this.searchMoviesUseCase = searchMoviesUseCase;
        this.getQueryHistoryUseCase = getQueryHistoryUseCase;
        this.saveQueryHistoryUseCase = saveQueryHistoryUseCase;
        this.clearQueryHistoryUseCase = clearQueryHistoryUseCase;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(SearchMoviesViewModel.class)) {
            return (T) new SearchMoviesViewModel(
                    searchMoviesUseCase,
                    getQueryHistoryUseCase,
                    saveQueryHistoryUseCase,
                    clearQueryHistoryUseCase);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: app/src/main/java/com/epam/valkaryne/openmovieapp/vm/ViewModelFactory.kt
package com.epam.valkaryne.openmovieapp.vm

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.epam.valkaryne.openmovieapp.domain.usecase.ClearQueryHistoryUseCase
import com.epam.valkaryne.openmovieapp.domain.usecase.GetQueryHistoryUseCase
import com.epam.valkaryne.openmovieapp.domain.usecase.SaveQueryHistoryUseCase
import com.epam.valkaryne.openmovieapp.domain.usecase.SearchMoviesUseCase

class ViewModelFactory(
    private val searchMoviesUseCase: SearchMoviesUseCase,
    private val getQueryHistoryUseCase: GetQueryHistoryUseCase,
    private val saveQueryHistoryUseCase: SaveQueryHistoryUseCase,
    private val clearQueryHistoryUseCase: ClearQueryHistoryUseCase
) : ViewModelProvider.Factory {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(SearchMoviesViewModel::class.java)) {
            return SearchMoviesViewModel(
                searchMoviesUseCase,
                getQueryHistoryUseCase,
                saveQueryHistoryUseCase,
                clearQueryHistoryUseCase
            ) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}